src/algebra/hom/centroid.lean

```diff
@@ -0,0 +1,239 @@
+  zero := 0,
+  zero_add := by intros; ext; apply zero_add,
+  add_zero := by intros; ext; apply add_zero,
+  add_comm := by intros; ext; apply add_comm }
+
+@[simp] lemma coe_zero : ⇑(0 : centroid_hom α) = 0 := rfl
+@[simp] lemma coe_add (f g : centroid_hom α) : ⇑(f + g) = f + g := rfl
+
+@[simp] lemma zero_apply (a : α) : (0 : centroid_hom α) a = 0 := rfl
+@[simp] lemma add_apply (f g : centroid_hom α) (a : α) : (f + g) a = f a + g a := rfl
+
+-- c.f https://github.com/leanprover-community/mathlib/blob/master/src/algebra/group/hom_instances.lean#L58
+instance : semiring (centroid_hom α) :=
+{ zero_mul := λ x, ext $ λ i, rfl,
+  mul_zero := λ x, ext $ λ i, add_monoid_hom.map_zero _,
+  left_distrib :=  λ x y z, ext $ λ i, add_monoid_hom.map_add _ _ _,
+  right_distrib := λ x y z, ext $ λ i, rfl,
+  .. centroid_hom.monoid,
+  .. centroid_hom.add_comm_monoid }
+
+lemma comp_mul_comm (T S : centroid_hom α) (a b : α) : (T ∘ S) (a * b) = (S ∘ T) (a * b) :=
+by rw [comp_app, map_mul_right, map_mul_left, ←map_mul_right, ←map_mul_left]
+
+end non_unital_non_assoc_semiring
+
+section non_unital_non_assoc_ring
+variables [non_unital_non_assoc_ring α]
+
+/-- Negation of `centroid_hom`s as a `centroid_hom`. -/
+instance : has_neg (centroid_hom α) :=
+⟨λ f, ⟨-f.to_add_monoid_hom,
+λ a b, by rw [add_monoid_hom.to_fun_eq_coe, add_monoid_hom.neg_apply, add_monoid_hom.neg_apply,
+  mul_neg, ← add_monoid_hom.to_fun_eq_coe, map_mul_left'],
+λ a b, by rw [add_monoid_hom.to_fun_eq_coe, add_monoid_hom.neg_apply, add_monoid_hom.neg_apply,
+  ← add_monoid_hom.to_fun_eq_coe, map_mul_right', ← neg_mul],
+⟩⟩
+
+instance : has_sub (centroid_hom α) :=
+⟨λ f g, ⟨f.to_add_monoid_hom - g.to_add_monoid_hom, λ a b, begin
+    change _ - _ = _,
+    convert (mul_sub _ _ _).symm,
+    { exact f.map_mul_left' _ _ },
+    { exact g.map_mul_left' _ _ }
+  end, λ a b, begin
+    change _ - _ = _,
+    convert (sub_mul _ _ _).symm,
+    { exact f.map_mul_right' _ _ },
+    { exact g.map_mul_right' _ _ }
+  end⟩⟩
+
+instance : add_comm_group (centroid_hom α) :=
+{ sub_eq_add_neg := λ _ _, ext $ λ _, sub_eq_add_neg _ _,
+  add_left_neg := λ _, ext $ λ _, add_left_neg _,
+  ..centroid_hom.add_comm_monoid, ..centroid_hom.has_neg, ..centroid_hom.has_sub }
+
+instance  : ring (centroid_hom α) :=
+{ ..centroid_hom.semiring, ..centroid_hom.add_comm_group }
+
+lemma coe_neg (f : centroid_hom α) : ⇑(-f) = -f := rfl
+lemma coe_sub (f g : centroid_hom α) : ⇑(f - g) = f - g := rfl
+
+lemma neg_apply (f : centroid_hom α) (a : α) : (-f) a = - f a := rfl
+lemma sub_apply (f g : centroid_hom α) (a : α) : (f - g) a = f a - g a := rfl
+
+end non_unital_non_assoc_ring
+
+section non_unital_ring
+variables [non_unital_ring α]
+
+/-- A prime associative ring has commutative centroid. -/
+@[reducible] -- See note [reducible non instances]
+def comm_ring (h: ∀ a b : α, (∀ r : α, a * r * b = 0) → a = 0 ∨ b = 0) :
+  comm_ring (centroid_hom α) :=
+{ mul_comm := λ f g, begin
```

Comment:
I think the convention is to stick the begin on the next line (I could be wrong about this)
```suggestion
{ mul_comm := λ f g,
  begin
```

Reply:
When declaring a term of a structure, not really, no.